urld829b80034_wjch111_PTSP2013Wang_tgz-pJ8-controllers_momcts_momctsCore_TestPlannerJ8/src/controllers/momcts/utils/SetOperation.java

```diff
@@ -0,0 +1,673 @@
+		Double v = map.get(k);
+		if(v == null) return 0.0;
+		else return (java.lang.Double) v;
+	}
+	
+	@SuppressWarnings("hiding")
+	public static <K,Double> int getIntValue(Map<K,Integer> map, K k){
+		Integer v = map.get(k);
+		if(v == null) return 0;
+		else return (java.lang.Integer) v;
+	}
+	
+	public static <T extends Comparable<T>> boolean isBetter(T a, T b, boolean maximize, boolean strict){
+		if( (maximize && a.compareTo(b) > 0) || (!maximize && a.compareTo(b)<0)){
+			return true;
+		}
+		if(!strict) return a.compareTo(b)==0;
+		return false;
+	}
+
+	public static <T extends Comparable<T>> boolean isBetter(T a, T b, boolean maximize){
+		return isBetter(a,b,maximize, true);
+	}
+	//---------Vector arithmatic Operators-----------
+	/**
+	 * Calculate addition of vectors a and b
+	 * @param a
+	 * @param b
+	 * @assure a and b should have the same size
+	 * @return
+	 */
+	@SuppressWarnings("hiding")
+	public static Vector<Double> adds(Vector<Double> as, Vector<Double> bs){
+		if(as.size()!=bs.size()) return null;
+		Vector<Double> addition = new Vector<Double>();
+		for(int i=0; i<as.size();i++){
+			BigDecimal a = new BigDecimal(as.get(i).toString());
+			BigDecimal b = new BigDecimal(bs.get(i).toString());
+			BigDecimal sum = a.add(b);
+			addition.add(java.lang.Double.valueOf(sum.toString()));//Double.valueOf(sum));
+		}
+		return addition;
+	}
+	
+	/**
+	 * Calculate the dot product between vector as and bs
+	 * @param as
+	 * @param bs
+	 * @assure a and b should have the same size
+	 * @return
+	 */
+	public static Vector<Double> muls(Vector<Double> as, Vector<Double> bs){
+		if(as.size()!=bs.size()) return null;
+		Vector<Double> multiplication = new Vector<Double>();
+		for(int i=0; i< as.size();i++){
+			BigDecimal a = new BigDecimal(as.get(i).toString());
+			BigDecimal b = new BigDecimal(bs.get(i).toString());
+			multiplication.add(java.lang.Double.valueOf(a.multiply(b).toString()));
+		}
+		return multiplication;
+	}
+	
+	public static Vector<Double> muls(Vector<Double> as, double b){
+		Vector<Double> multiplication = new Vector<Double>();
+		for(int i=0; i< as.size();i++){
+			multiplication.add(as.get(i)*b);
+		}
+		return multiplication;
+	}
+	
+	public static Vector<Double> divs(Vector<Double> as, double b){
+		Vector<Double> multiplication = new Vector<Double>();
+		for(int i=0; i< as.size();i++){
+			multiplication.add(as.get(i)/b);
+		}
+		return multiplication;
+	}
+	
+	//---------end of Vector arithmatic Operators-----------
+	
+	//---------Map arithmatic Operators-----------
+	public static <T> HashMap<T, Double> muls(Map<T, Double> mp, double b){
+		HashMap<T, Double> mulMp = new HashMap<T, Double>();
+		for(T k : mulMp.keySet()){
+			mulMp.put(k, mp.get(k)*b);
+		}
+		return mulMp;
+	}
+
+	public static <T> HashMap<T, Double> divs(Map<T, Double> mp, double b){
+		HashMap<T, Double> divMp = new HashMap<T, Double>();
+		for(T k : mp.keySet()){
+			divMp.put(k, Presentation.ndigits(mp.get(k)/b,3));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.